src/Microsoft.AspNetCore.Mvc.Core/Routing/EndpointRoutingUrlHelper.cs

```diff
@@ -51,46 +52,41 @@ public override string Action(UrlActionContext urlActionContext)
                 throw new ArgumentNullException(nameof(urlActionContext));
             }
 
-            var valuesDictionary = GetValuesDictionary(urlActionContext.Values);
+            var values = GetValuesDictionary(urlActionContext.Values);
 
             if (urlActionContext.Action == null)
             {
-                if (!valuesDictionary.ContainsKey("action") &&
+                if (!values.ContainsKey("action") &&
                     AmbientValues.TryGetValue("action", out var action))
                 {
-                    valuesDictionary["action"] = action;
+                    values["action"] = action;
                 }
             }
             else
             {
-                valuesDictionary["action"] = urlActionContext.Action;
+                values["action"] = urlActionContext.Action;
             }
 
             if (urlActionContext.Controller == null)
             {
-                if (!valuesDictionary.ContainsKey("controller") &&
+                if (!values.ContainsKey("controller") &&
                     AmbientValues.TryGetValue("controller", out var controller))
                 {
-                    valuesDictionary["controller"] = controller;
+                    values["controller"] = controller;
                 }
             }
             else
             {
-                valuesDictionary["controller"] = urlActionContext.Controller;
+                values["controller"] = urlActionContext.Controller;
             }
 
-            var successfullyGeneratedLink = _linkGenerator.TryGetLink(
-                ActionContext.HttpContext,
-                valuesDictionary,
-                out var link);
-            if (!successfullyGeneratedLink)
-            {
-                //TODO: log here
 
-                return null;
-            }
-
-            return GenerateUrl(urlActionContext.Protocol, urlActionContext.Host, link, urlActionContext.Fragment);
+            var path = _linkGenerator.GetPathByRouteValues(
```

Comment:
What's the API to determine if a link was successfully generated (for https://github.com/aspnet/Mvc/issues/8287)?

Reply:
I believe null is returned for failure

Reply:
So `null` -> no route, empty string -> valid link to route at root. `TryGet` had crispier semantics, but we can live with this.

Reply:
LinkGenerator will return `/` for a successful root link. It's a path, not a virtual path. 

We started with Get and TryGet but there were a few issues.
1. If Get throws when it fails, it just feels really bad. 
2. If get doesn't throw, then tryGet isn't needed
3. We ended up with 8 methods multiplied by a number of different sets of APIs. It feel really overwhelming without much value. 

We resolved to add an additional flavour of APIs taht throw if there's demand for it. Like `GetRequriredPath...`. We could do this at any point since it's additive.